main.go

```diff
@@ -449,8 +435,32 @@ func (a *arg) list(proxy string) error {
 	return nil
 }
 
-func debugf(format string, args ...interface{}) {
-	if debug || *fDebug {
-		fmt.Fprintf(os.Stderr, format, args...)
+func run(cmd *exec.Cmd) error {
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	start := time.Now()
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run %v: %v\n%s", strings.Join(cmd.Args, " "), err, stderr.String())
+	}
+
+	end := time.Now()
+
+	if !debug && !*fDebug {
+		return nil
+	}
+
+	env := cmd.Env
+	if env == nil {
+		env = os.Environ()
+	}
+	var goenv []string
+	for _, v := range env {
+		if strings.HasPrefix(v, "GO") {
+			goenv = append(goenv, v)
+		}
 	}
+	fmt.Fprintf(os.Stderr, "ran [%v %v] with Dir [%v] in %v\n%s", strings.Join(goenv, " "), strings.Join(cmd.Args, " "), cmd.Dir, end.Sub(start), stderr.String())
```

Comment:
FWIW I've found it quite nice when debugging output like this prints the command in a form that is suitable for input for the shell.


    cd /some/dir
    GOPATH=/foo/bar acommand args 'with appropriate quoting'
 
might be nice to distinguish stderr from commands too. maybe with a leading "+ " on the commands?

Reply:
Good point on the `cd`. I'd sorted the env variables but not gone the full distance. Great suggestion, thanks